File: library/src/main/java/com/pengrad/telegrambot/model/request/InlineQueryResult.java
package com.pengrad.telegrambot.model.request;

import com.pengrad.telegrambot.model.MessageEntity;

import java.io.Serializable;

/**
 * stas
 * 1/12/16.
 */
public abstract class InlineQueryResult<T extends InlineQueryResult<T>> implements Serializable {
    private final static long serialVersionUID = 0L;

    @SuppressWarnings("unchecked")
    private final T thisAsT = (T) this;

    private String type;
    private String id;
    private MessageEntity[] caption_entities;
    private InputMessageContent input_message_content;
    private InlineKeyboardMarkup reply_markup;
    private Boolean show_caption_above_media;

    public InlineQueryResult(String type, String id) {
        this.type = type;
        this.id = id;
    }

    public T captionEntities(MessageEntity... entities) {
        this.caption_entities = entities;
        return thisAsT;
    }

    public T inputMessageContent(InputMessageContent inputMessageContent) {
        this.input_message_content = inputMessageContent;
        return thisAsT;
    }

    public T replyMarkup(InlineKeyboardMarkup replyMarkup) {
        this.reply_markup = replyMarkup;
        return thisAsT;
    }

    public T showCaptionAboveMedia(Boolean showCaptionAboveMedia) {
        this.show_caption_above_media = showCaptionAboveMedia;
        return thisAsT;
    }

}

File: core/src/main/kotlin/com/pengrad/telegrambot/model/request/InlineQueryResult.kt
package com.pengrad.telegrambot.model.request

import java.io.Serializable

/**
 * stas
 * 1/12/16.
 */
abstract class InlineQueryResult<T>(
        private val type: String,
        private val id: String
) : Serializable {

    private val thisAsT = this as T
    private var input_message_content: InputMessageContent? = null
    private var reply_markup: InlineKeyboardMarkup? = null

    fun inputMessageContent(inputMessageContent: InputMessageContent): T {
        this.input_message_content = inputMessageContent
        return thisAsT
    }

    fun replyMarkup(replyMarkup: InlineKeyboardMarkup): T {
        this.reply_markup = replyMarkup
        return thisAsT
    }

    companion object {
        private const val serialVersionUID = 0L
    }
}
